docs/src/components/BuiltinSearch/index.js

```diff
@@ -0,0 +1,164 @@
+import builtins from "@generated/builtin-data/default/builtins.json";
+import React, { useMemo, useState } from "react";
+import ReactMarkdown from "react-markdown";
+import styles from "./styles.module.css";
+
+export default function BuiltinSearch({ entryLimit }) {
+  const [searchTerm, setSearchTerm] = useState("");
+  const [selectedCategory, setSelectedCategory] = useState("");
+  const [isWasm, setIsWasm] = useState(false);
+  const displayLimit = entryLimit;
+  const allRows = useMemo(() => {
+    return (Object.keys(builtins._categories)
+      .map((category) => {
+        return builtins._categories[category].map((builtin) => {
+          const fn = builtins[builtin];
+          if (!fn) return null;
+
+          const anchor = `builtin-${category}-${builtin.replaceAll(".", "")}`;
+
+          const isInfix = !!fn.infix;
+          const isRelation = !!fn.relation;
+
+          const args = fn.args || [];
+          const result = fn.result || {};
+
+          const signature = isInfix
+            ? `${args[0]?.name || "x"} ${fn.infix} ${args[1]?.name || "y"}`
+            : isRelation
+            ? `${builtin}(${args.map((a) => a.name).join(", ")}, ${result.name})`
+            : `${result.name || "result"} := ${builtin}(${args.map((a) => a.name).join(", ")})`;
+          return {
+            name: builtin,
+            category: category,
+            wasm: fn.wasm,
+            introduced: fn.introduced,
+            infix: isInfix,
+            signature: signature,
+            anchor: anchor,
+            versions: fn.available,
+            deprecated: fn?.deprecated ?? false,
+            internal: category == "internal",
+          };
+        });
+      })).reduce((a, b) => a.concat(b), []);
+  });
+  const filteredRows = allRows.map((row) => {
+    return ((row.name.includes(searchTerm) || row.signature.includes(searchTerm) && row.infix)
+        && (!selectedCategory || row.category == selectedCategory) && (!isWasm || row.wasm))
+      ? row
+      : null;
+  }).filter((a) => a != null && !(a.internal || a.deprecated) && (searchTerm || selectedCategory || isWasm));
```

Comment:
It might be easier to read if you had multiple filters, one for category, one for name, one for isWasm etc. Then it'd be built up more incrementally and I think easier to understand. Even if it's slower to filter the list more than once, each time the number of items being filtered is smaller.

Reply:
Since this runs every time the search term changes I think it's good to avoid adding too many extra steps, decided to add comments to each step of the statement